flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientSideSerializationTest.java

```diff
@@ -0,0 +1,221 @@
+		handler.addInputChannel(inputChannel);
+
+		channel = new EmbeddedChannel(
+			new NettyMessage.NettyMessageEncoder(), // For outbound messages
+			new NettyMessageClientDecoderDelegate(handler)); // For inbound messages
+
+		inputChannelId = inputChannel.getInputChannelId();
+	}
+
+	@After
+	public void tearDown() throws IOException {
+		inputGate.close();
+
+		networkBufferPool.destroyAllBufferPools();
+		networkBufferPool.destroy();
+
+		channel.close();
+	}
+
+	@Test
+	public void testBufferResponse() {
+		NetworkBuffer buffer = new NetworkBuffer(MemorySegmentFactory.allocateUnpooledSegment(BUFFER_SIZE), FreeingBufferRecycler.INSTANCE);
+
+		for (int i = 0; i < BUFFER_SIZE; i += 8) {
+			buffer.writeLong(i);
+		}
+
+		Buffer testBuffer = testReadOnlyBuffer ? buffer.readOnlySlice() : buffer;
+		if (testCompressedBuffer) {
+			testBuffer = COMPRESSOR.compressToOriginalBuffer(buffer);
+		}
+
+		NettyMessage.BufferResponse expected = new NettyMessage.BufferResponse(
+			testBuffer, random.nextInt(), inputChannelId, random.nextInt());
+		NettyMessage.BufferResponse actual = encodeAndDecode(expected, channel);
+
+		assertTrue(buffer.isRecycled());
+		assertTrue(testBuffer.isRecycled());
+
+		assertNotNull(
+			"The request input channel should always have available buffers in this test.",
+			actual.getBuffer());
+		ByteBuf retainedSlice = actual.getBuffer().asByteBuf();
+		if (testCompressedBuffer) {
+			assertTrue(actual.isCompressed);
+			retainedSlice = decompress(retainedSlice);
+		}
+
+		// Ensure not recycled and same size as original buffer
+		assertEquals(1, retainedSlice.refCnt());
+		assertEquals(BUFFER_SIZE, retainedSlice.readableBytes());
+
+		for (int i = 0; i < BUFFER_SIZE; i += 8) {
+			assertEquals(i, retainedSlice.readLong());
+		}
+
+		// Release the received message.
+		actual.releaseBuffer();
+		if (testCompressedBuffer) {
+			retainedSlice.release();
+		}
+		assertEquals(0, retainedSlice.refCnt());
+
+		verifyBufferResponseHeader(expected, actual);
+	}
+
+	@Test
+	public void testErrorResponseWithoutErrorMessage() {
+		IllegalStateException expectedError = new IllegalStateException();
+		InputChannelID receiverId = new InputChannelID();
+
+		NettyMessage.ErrorResponse expected = new NettyMessage.ErrorResponse(expectedError, receiverId);
+		NettyMessage.ErrorResponse actual = encodeAndDecode(expected, channel);
+
+		verifyErrorResponse(expected, actual);
+	}
+
+	@Test
+	public void testErrorResponseWithErrorMessage() {
+		IllegalStateException expectedError = new IllegalStateException("Illegal illegal illegal");
+		InputChannelID receiverId = new InputChannelID();
+
+		NettyMessage.ErrorResponse expected = new NettyMessage.ErrorResponse(expectedError, receiverId);
+		NettyMessage.ErrorResponse actual = encodeAndDecode(expected, channel);
+
+		verifyErrorResponse(expected, actual);
+	}
+
+	@Test
+	public void testErrorResponseWithFatalError() {
```

Comment:
Deduplicate the codes for above `testErrorResponseWithFatalError`, `testErrorResponseWithErrorMessage` and `testErrorResponseWithoutErrorMessage`.

Reply:
Extractor the common method.